### DISCUSSION TITLE
插入排序法

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#106] [其他問題] [answered]
參考資料 : 
1. https://blog.techbridge.cc/2017/08/19/sotring-algorithm/
2. https://zh.wikipedia.org/wiki/%E6%8F%92%E5%85%A5%E6%8E%92%E5%BA%8F

參考胡立老師的文章和維基百科，有兩個地方有點疑惑
1. 老師的內層迴圈，while 部分，因為是去找前面已經排序好的位置，所以應該是 position -1，如果是 i >=0 這個條件永遠都會成立，作為迴圈的終止條件之一，好像有點怪怪的
2. 不管是老師的文章，還是維基百科，都會在最後外層迴圈加上 `arr[position] = value`，可是我在內層迴圈就已經將 arr 裡的元素兩兩位置互換，所以`arr[position] = value` 這段程式碼是不是可以省略 ? 

``` js
const Sort = (arr) => {
  // 假設第一個元素已經排好，所以從 1 開始跑 
  for (let i = 1; i < arr.length ; i++) {

    let position = i // 要插入元素的位置
	const value = arr[i] // 要插入的元素值
    
    while (position-1 >= 0 && arr[position-1] > value) {
      let temp = arr[position]
      arr[position] = arr[position-1]
      arr[position-1] = temp
      // [arr[position], arr[position - 1]] = [arr[position - 1], arr[position]]
      position --

    }
    // arr[position] = value
  }
	return arr
}

module.exports = Sort
// [1, 3, 0, 6 , 8, 7, 2, 4]
// 已排好 [1] 未排好 [3, 0, 6 , 8, 7, 2, 4]
// for : i=1; 1<8 => position = 1 , value = 3 
// while :從已排好的元素中從後向前掃描 0 >=0 ,  arr[0]=1 沒有大於 3 => while 結束=> 更新位置 =>arr[position] = arr[1] = 3  

// 已排好 [1, 3] 未排好 [ 0, 6 , 8, 7, 2, 4]
// for : i=2; 2<8 => position = 2 , value = 0
// while :從已排好的元素中從後向前掃描 
// 條件 : 1 >=0 ,  arr[1]=3 大於 0 => while 執行
// 互換位置 => [1,0,3] position = 1
// 條件 : 0 >=0 , arr[0]=1 大於 0 => while 執行
// 互換位置 => [0,1,3] position = 0 
```

**Comment by allenliaooooooo** [comment, marked as answer]
> 1. 老師的內層迴圈，while 部分，因為是去找前面已經排序好的位置，所以應該是 position -1，如果是 i >=0 這個條件永遠都會成立，作為迴圈的終止條件之一，好像有點怪怪的

同意~
> 2. 不管是老師的文章，還是維基百科，都會在最後外層迴圈加上 `arr[position] = value`，可是我在內層迴圈就已經將 arr 裡的元素兩兩位置互換，所以`arr[position] = value` 這段程式碼是不是可以省略 ?

我覺得可以~~
維基會在迴圈外層加上 `arr[position] = value` 是因為他沒有把位置互換，只是把上一個位置的值賦給現在這個位置的值
但上一個位置的值並沒有更動到

意外發現老師的文章網址英文好像打錯了 XD

**Reply by s103071049** [reply]
真的也 XDD  感謝~

**Reply by aszx87410** [reply]
沒錯，你們是對的
